fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexBuildState.java

```
@@ -0,0 +1,160 @@
+ * </p>
+ * <p>
+ * Note that index build state (especially records scanned) serves a informational purpose only in terms that records scanned is not necessarily 100% of records in total when the index build completes in several reasons:
+ * </p>
+ * <ul>
+ *     <li> Records scanned does not include the records scanned in the {@link OnlineIndexer} whose {@link OnlineIndexer.Builder#setTrackProgress(boolean)} is set to false. </li>
+ *     <li> Records added or deleted during the index build progress are not included in records scanned but are in records in total. </li>
+ *     <li> Records in total uses the number of all records in the store. In some cases, the real number of records that need to be scanned is less than this.</li>
+ * </ul>
+ * @see #getIndexBuildStateAsync(FDBRecordStore, Index)
+ */
+@API(API.Status.UNSTABLE)
+public class IndexBuildState {
+    @Nonnull
+    private final IndexState indexState;
+    @Nullable
+    private final Long recordsScanned;
+    @Nullable
+    private final Long recordsInTotal;
+
+    /**
+     * Get the build progress ({@link IndexBuildState}) of the given index in the given record store asynchronously.
+     * @param store the record store containing the index
+     * @param index the index needed to be checked
+     * @return a future that completes to the index build state
+     */
+    public static CompletableFuture<IndexBuildState> getIndexBuildStateAsync(FDBRecordStore store, Index index) {
+        IndexState indexState = store.getIndexState(index);
+        if (indexState != IndexState.WRITE_ONLY) {
+            return CompletableFuture.completedFuture(new IndexBuildState(indexState));
+        }
+        return getRecordsScannedAsync(store, index).thenCompose(scannedRecords -> {
+            try {
+                return store.getSnapshotRecordCount()
+                        .thenApply(totalRecords -> new IndexBuildState(indexState, scannedRecords, totalRecords));
+            } catch (Exception ex) {
```

This should probably catch `RecordCoreException`. It's also possible that we should introduce a new exception type, like, `AggregateFunctionNotSupported` or something, and that that function should throw that error, but catching `RecordCoreException` is probably fine for now. (And the exception can be changed safely if `AggregateFunctionNotSupported` is a sub-type of `RecordCoreException`.)